apollo-router/src/orbiter/mod.rs

```diff
@@ -0,0 +1,416 @@
+    tracing::debug!("anonymous usage: {}", serde_json::to_string_pretty(&body)?);
+
+    #[cfg(not(test))]
+    let url = "https://router.apollo.dev/telemetry";
+    #[cfg(test)]
+    let url = "http://localhost:8888/telemetry";
+
+    Ok(reqwest::blocking::Client::new()
+        .post(url)
+        .header(USER_AGENT, "router")
+        .header(CONTENT_TYPE, "application/json")
+        .json(&serde_json::to_value(body)?)
+        .timeout(Duration::from_secs(10))
+        .send()?
+        .text()?)
+}
+
+fn get_os() -> String {
+    if wsl::is_wsl() {
+        "wsl"
+    } else {
+        std::env::consts::OS
+    }
+    .to_string()
+}
+
+fn visit_config(usage: &mut HashMap<String, u64>, config: &Value) {
+    // We have to be careful not to expose names of headers, metadata or anything else sensitive.
+    let raw_json_schema =
+        serde_json::to_value(generate_config_schema()).expect("config schema is must be valid");
+    let compiled_json_schema = JSONSchema::compile(
+        &serde_json::to_value(&raw_json_schema).expect("config schema is must be valid"),
+    )
+    .expect("config schema must compile");
+
+    // We can use jsonschema to give us annotations about the validated config. This means that we get
+    // a pointer into the config document and also a pointer into the schema.
+    // For this to work ALL config must have an annotation, e.g. documentation.
+    // For each config path we need to sanitize the it for arrays and also custom names e.g. header names.
+    // This corresponds to the json schema keywords of `items` and `additionalProperties`
+    if let BasicOutput::Valid(output) = compiled_json_schema.apply(config).basic() {
+        for item in output {
+            let instance_ptr = item.instance_location();
+            let value = config
+                .pointer(&instance_ptr.to_string())
+                .expect("pointer must point to value");
+
+            // Compose the redacted path.
+            let mut path = Vec::new();
+            for chunk in item.keyword_location() {
+                if let PathChunk::Property(property) = chunk {
+                    // We hit a properties keyword, we can grab the next keyword as it'll be a property name.
+                    path.push(property.to_string());
+                }
+                if &PathChunk::Keyword("additionalProperties") == chunk {
+                    // This is free format properties. It's redacted
+                    path.push("redacted".to_string());
+                }
+            }
+
+            let path = path.join(".");
+            if matches!(item.keyword_location().last(), Some(&PathChunk::Index(_))) {
+                *usage
+                    .entry(format!("configuration{}.len", path))
+                    .or_default() += 1;
+            }
+            match value {
+                Value::Bool(value) => {
+                    *usage
+                        .entry(format!("configuration{}.{}", path, value))
```

Comment:
maybe add a `.` between `configuration` and the option name? on the other ones too

Reply:
It does end up that way as the path always begins with a dot. The path into the instance always begins with a slash, and we replace them with dots.

Reply:
looking at the snapshots, some of them have no dot: https://github.com/apollographql/router/pull/2173/files#diff-8a5c9f4b77740fe9029fccc0a42dd45ed7511c3413fe4fa177fed5f878d89454R12-R26

Reply:
You are right.

Reply:
[c1e86c2](https://github.com/apollographql/router/pull/2173/commits/c1e86c22a7aacc304202d30108bfc28a6a9d60dc)